### ISSUE TITLE
朋友们, 你们好. 我想请问关于attention计算中mask的问题.

### ISSUE THREAD

**Issue by cuiruochu** [jingyaogong/minimind#448]
对于模型的输入X, X中似乎含有用于padding的token, 在本项目中对于的token id 为 0.

在预训练的过程中, 对于Attention的计算, 用于padding的token是不是应该mask掉呢? 

在MiniMindForCausalLM类的前向传播中有attention_mask参数, 但是在预训练时似乎设置为了None.

所以我有疑问, 在训练时应该mask掉那些用于padding的token吗?

**Comment by dao86** [comment]
用于padding的token，一般设为0。但我想，你应该说的是被padding的token吧？

**Comment by cuiruochu** [comment]
> 用于padding的token，一般设为0。但我想，你应该说的是被padding的token吧？

朋友你好，我指的就是用于padding的token，在minimind的tokenizer中，就是id为0的token。这个id为0的token用于在训练时对输入进行padding。
我的问题是，这个token似乎不应该参与Attention的计算，在minimind的源码中也有考虑到这一点（forward时存在attention_mask参数）。
但是这个attention_mask在预训练时被设置为None，导致Attention的计算并没有屏蔽掉这些token。这对于模型的最终效果有影响吗？

**Comment by dao86** [comment]
计算注意力时，需要mask的，一般有两种情况，一种是训练时的未来token,这部分在注意力代码中会自行计算(scores+mask)，不需要外部传入。   
另一种是，用户在外部根据其他情况，传入的attention_mask。一般外面传入的attention_mask，是用户自己定义的一些掩码，不是指正常预训练时的自注意力未来掩码。所以你需要先弄明白，哪些token是需要被mask的。
这是mini中attention的代码，请注意scores+mask 和  attention_mask的区别
scores = (xq @ xk.transpose(-2, -1)) / math.sqrt(self.head_dim)
            scores = scores + torch.triu(
                torch.full((seq_len, seq_len), float("-inf"), device=scores.device),
                diagonal=1
            ).unsqueeze(0).unsqueeze(0)  **# scores+mask**

            if attention_mask is not None:
                extended_attention_mask = attention_mask.unsqueeze(1).unsqueeze(2)
                extended_attention_mask = (1.0 - extended_attention_mask) * -1e9
                scores = scores + extended_attention_mask

**Comment by Cui-ruochu** [comment]
> 计算注意力时，需要mask的，一般有两种情况，一种是训练时的未来token,这部分在注意力代码中会自行计算(scores+mask)，不需要外部传入。 另一种是，用户在外部根据其他情况，传入的attention_mask。一般外面传入的attention_mask，是用户自己定义的一些掩码，不是指正常预训练时的自注意力未来掩码。所以你需要先弄明白，哪些token是需要被mask的。 这是mini中attention的代码，请注意scores+mask 和 attention_mask的区别 scores = (xq @ xk.transpose(-2, -1)) / math.sqrt(self.head_dim) scores = scores + torch.triu( torch.full((seq_len, seq_len), float("-inf"), device=scores.device), diagonal=1 ).unsqueeze(0).unsqueeze(0) **# scores+mask**
> 
> ```
>         if attention_mask is not None:
>             extended_attention_mask = attention_mask.unsqueeze(1).unsqueeze(2)
>             extended_attention_mask = (1.0 - extended_attention_mask) * -1e9
>             scores = scores + extended_attention_mask
> ```

对的朋友，我想问就是你上面所说的：用户在外部根据其他情况，传入的attention_mask。

在训练时，假设input为[ [1, 1235, 2352, 6574, 1, 1, 1, 1 ,1 ], ]。在input中有5个用于padding的token。
我在计算Attention时，是不是应该同时使用attention_mask来保证这些id为1的token不参与计算。

**Comment by cuiruochu** [comment]
> > 计算注意力时，需要mask的，一般有两种情况，一种是训练时的未来token,这部分在注意力代码中会自行计算(scores+mask)，不需要外部传入。 另一种是，用户在外部根据其他情况，传入的attention_mask。一般外面传入的attention_mask，是用户自己定义的一些掩码，不是指正常预训练时的自注意力未来掩码。所以你需要先弄明白，哪些token是需要被mask的。 这是mini中attention的代码，请注意scores+mask 和 attention_mask的区别 scores = (xq @ xk.transpose(-2, -1)) / math.sqrt(self.head_dim) scores = scores + torch.triu( torch.full((seq_len, seq_len), float("-inf"), device=scores.device), diagonal=1 ).unsqueeze(0).unsqueeze(0) **# scores+mask**
> > ```
> >         if attention_mask is not None:
> >             extended_attention_mask = attention_mask.unsqueeze(1).unsqueeze(2)
> >             extended_attention_mask = (1.0 - extended_attention_mask) * -1e9
> >             scores = scores + extended_attention_mask
> > ```
> 
 对的朋友，我想问就是你上面所说的：用户在外部根据其他情况，传入的attention_mask。

在训练时，假设input为[ [1, 1235, 2352, 6574, 1, 1, 1, 1 ,1 ], ]。在input中有5个用于padding的token。 我在计算Attention时，是不是应该同时使用attention_mask来保证这些id为1的token不参与计算。

if attention_mask is not None:
        extended_attention_mask = attention_mask.unsqueeze(1).unsqueeze(2)
        extended_attention_mask = (1.0 - extended_attention_mask) * -1e9
        scores = scores + extended_attention_mask

**Comment by dao86** [comment]
是的，你如果不希望后面的数值为1的token参与注意力计算，就需要使用attention_mask。但需要注意，根据代码attention_mask和scores 的形状要相同

**Comment by cuiruochu** [comment]
> 是的，你如果不希望后面的数值为1的token参与注意力计算，就需要使用attention_mask。但需要注意，根据代码attention_mask和scores 的形状要相同

感谢你的解答。
